pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go

```diff
@@ -385,3 +392,83 @@ func (ca *cloudCIDRAllocator) ReleaseCIDR(node *v1.Node) error {
 		node.Name, node.Spec.PodCIDR)
 	return nil
 }
+
+func (ca *cloudCIDRAllocator) updateMultiNetworkAnnotations(node *v1.Node, northInterfaces networkv1.NorthInterfacesAnnotation, additionalNodeNetworks networkv1.MultiNetworkAnnotation) error {
+	northInterfaceAnn, err := networkv1.MarshalNorthInterfacesAnnotation(northInterfaces)
+	if err != nil {
+		klog.ErrorS(err, "Failed to marshal the north interfaces annotation for multi-networking", "nodeName", node.Name)
+		return err
+	}
+	additionalNodeNwAnn, err := networkv1.MarshalAnnotation(additionalNodeNetworks)
+	if err != nil {
+		klog.ErrorS(err, "Failed to marshal the additional node networks annotation for multi-networking", "nodeName", node.Name)
+		return err
+	}
+	if node.Annotations == nil {
+		node.Annotations = make(map[string]string)
+	}
+	node.Annotations[networkv1.NorthInterfacesAnnotationKey] = northInterfaceAnn
+	node.Annotations[networkv1.MultiNetworkAnnotationKey] = additionalNodeNwAnn
+	node.Status.Capacity, err = allocateIPCapacity(node, additionalNodeNetworks)
+	if err != nil {
+		return err
+	}
+	// Prepare patch bytes for the node update.
+	patchBytes, err := json.Marshal([]interface{}{
+		map[string]interface{}{
+			"op":    "replace",
+			"path":  "/metadata/annotations",
+			"value": node.Annotations,
+		},
+		map[string]interface{}{
+			"op":    "add",
+			"path":  "/status/capacity",
+			"value": node.Status.Capacity,
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("failed to build patch bytes for multi-networking: %v", err)
+	}
+	// Since dynamic network addition/deletion is a use case to be supported, we aspire to build these annotations and IP capacities every time from scratch.
+	// Hence we are using a JSON patch merge strategy instead of strategic merge strategy on the node during update.
+	if _, err = ca.client.CoreV1().Nodes().Patch(context.TODO(), node.Name, types.JSONPatchType, patchBytes, metav1.PatchOptions{}, "status"); err != nil {
+		nodeutil.RecordNodeStatusChange(ca.recorder, node, "CIDRAssignmentFailed")
+		klog.ErrorS(err, "Failed to update the node annotations and capacity for multi-networking", "nodeName", node.Name)
+		return err
+	}
+	return nil
+}
+
+// allocateIPCapacity updates the extended IP resource capacity for every non-default network on the node.
+func allocateIPCapacity(node *v1.Node, nodeNetworks networkv1.MultiNetworkAnnotation) (v1.ResourceList, error) {
```

Comment:
Would be good to have test code for this function. It should exercise all different code paths.

Reply:
We would have to export that function to be able to write tests for it. I've added additional tests for `updateMultiNetworkAnnotations` that cover `allocateIPCapacity` as well.